cirq/contrib/quimb/mps_simulator.py

```diff
@@ -231,92 +235,243 @@ def sample(
+                    pop_idx = idx
+                elif smallest_badness > badness:
+                    smallest_badness = badness
+                    pop_idx = idx
+
+            # Pop the element to aggregate
+            qubit = qubit_queue.pop(pop_idx)
+
+            i = self.qubit_map[qubit]
+            Mi = self.M[i]
+            if _trace_out(i):
+                Mi = _sum_reduce(Mi, self.i_str(i))
+            M = _make_tensor_if_scalar(M @ Mi)
+        return M
+
+    def partial_state_vector(self, skip_tracing_out_for_qubits):
+        M = self._sum_up(
+            skip_tracing_out_for_qubits={
+                self.qubit_map[qubit] for qubit in skip_tracing_out_for_qubits
+            }
+        )
+        # Here, we rely on the formatting of the indices, and the fact that we have enough
+        # leading zeros so that 003 comes before 100.
+        sorted_ind = tuple(sorted(M.inds))
+        return M.fuse({'i': sorted_ind}).data
+
     def state_vector(self):
-        M = np.ones((1, 1))
-        for i in range(len(self.M)):
-            M = np.einsum('ni,npj->pij', M, self.M[i])
-            M = M.reshape(M.shape[0], -1)
-        assert M.shape[0] == 1
-        return M[0, :]
+        return self.partial_state_vector(skip_tracing_out_for_qubits={})
 
     def to_numpy(self) -> np.ndarray:
         return self.state_vector()
 
     def apply_unitary(self, op: 'cirq.Operation'):
-        idx = [self.qubit_map[qubit] for qubit in op.qubits]
         U = protocols.unitary(op).reshape([qubit.dimension for qubit in op.qubits] * 2)
 
-        if len(idx) == 1:
-            n = idx[0]
-            self.M[n] = np.einsum('ij,mnj->mni', U, self.M[n])
-        elif len(idx) == 2:
-            if abs(idx[0] - idx[1]) != 1:
-                raise ValueError('Can only handle continguous qubits')
-            elif idx[0] < idx[1]:
-                n, p = idx
-            else:
-                p, n = idx
-                U = np.swapaxes(np.swapaxes(U, 0, 1), 2, 3)
-            T = np.einsum('klij,mni,npj->mkpl', U, self.M[n], self.M[p])
-            X, S, Y = np.linalg.svd(T.reshape([T.shape[0] * T.shape[1], T.shape[2] * T.shape[3]]))
-            X = X.reshape([T.shape[0], T.shape[1], -1])
-            Y = Y.reshape([-1, T.shape[2], T.shape[3]])
-
-            S = np.asarray([math.sqrt(x) for x in S])
-
-            nkeep = 0
-            for i in range(S.shape[0]):
-                if S[i] >= S[0] * self.threshold:
-                    nkeep = i + 1
-
-            X = X[:, :, :nkeep]
-            S = np.diag(S[:nkeep])
-            Y = Y[:nkeep, :, :]
-
-            self.M[n] = np.einsum('mis,sn->mni', X, S)
-            self.M[p] = np.einsum('ns,spj->npj', S, Y)
+        if len(op.qubits) == 1:
+            n = self.qubit_map[op.qubits[0]]
+
+            old_n = self.i_str(n)
+            new_n = 'new_' + old_n
+
+            U = qtn.Tensor(U, inds=(new_n, old_n))
+            self.M[n] = (U @ self.M[n]).reindex({new_n: old_n})
```

Comment:
so you're actually contracting in the one qubit gates here explicitly? I think usually with quimb you can just throw everything into its own `Tensor` and contract in one fell swoop at the end, but maybe to schedule the SVD's how you want, you have to do it this way

Reply:
How would you do this? Here's what I tried:

```
tensor_network
(Tensor(shape=(2,), inds=('i_0',), tags=set()), Tensor(shape=(2,), inds=('i_1',), tags=set()))

U
Tensor(shape=(2, 2), inds=('new_i_0', 'i_0'), tags=set())

new_tensor = tensor_network @ U
Tensor(shape=(2, 2), inds=('i_1', 'new_i_0'), tags=set())

new_network = tensor_network & U
(Tensor(shape=(2,), inds=('i_0',), tags=set()), Tensor(shape=(2,), inds=('i_1',), tags=set()), Tensor(shape=(2, 2), inds=('new_i_0', 'i_0'), tags=set()))

```
So if I use the `@` operator I go from a network with 2 tensors to a single tensor, and if I use the `&` operator, then I get non-collapsed tensors.

Leaving unresolved because I suspect I am missing something.

Reply:
The key point is that it contracts any indices that share the same name. In your example, there is still a tensor-factorized structure between the first and second tensor in `tensor_network` and `new_network`. 

![File_000](https://user-images.githubusercontent.com/4967059/105563043-9012bf00-5cd1-11eb-8e17-a3be49a641a2.jpeg)

Reply:
optional for current PR, since I think it would be a whole exercise to try to re-write this function to wholely use quimb's contraction engine. I'm pleased you got it working in `_sum_up`.

Reply:
Ack. Added a TODO.